src/main/java/com/google/devtools/build/lib/sandbox/cgroups/VirtualCGroup.java

```diff
@@ -0,0 +1,208 @@
+package com.google.devtools.build.lib.sandbox.cgroups;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Streams;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.io.CharSink;
+import com.google.common.io.FileWriteMode;
+import com.google.common.io.Files;
+import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.EventHandler;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.Controller;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v1.LegacyCpu;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v1.LegacyMemory;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v2.UnifiedCpu;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v2.UnifiedMemory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Scanner;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+
+/**
+ * This class creates and exposes a virtual cgroup for the bazel process and allows creating
+ * child cgroups. Resources are exposed as {@link Controller}s, each representing a
+ * subsystem within the virtual cgroup and that could in theory belong to different real cgroups.
+ */
+@AutoValue
+public abstract class VirtualCGroup {
+    private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+    private static final File PROC_SELF_MOUNTS_PATH = new File("/proc/self/mounts");
+    private static final File PROC_SELF_CGROUP_PATH = new File("/proc/self/cgroup");
+
+    @Nullable
+    private static volatile VirtualCGroup instance;
+
+    @Nullable
+    public abstract Controller.Cpu cpu();
+    @Nullable
+    public abstract Controller.Memory memory();
+    public abstract ImmutableSet<Path> paths();
```

Comment:
Would it be easier to work with `File(s)` (https://docs.oracle.com/javase/8/docs/api/java/io/File.html) here? We seem to be calling `Path#toFile()` quite often.